File: Data Structures & Algorithms/maximum-product-subarray/submission-0.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:

        maxProduct = nums[0]

        if len(nums) == 1:
            return maxProduct
        
        for i in range(len(nums)):
            currProduct = nums[i]
            maxProduct = max(maxProduct, currProduct)
            for j in range(i+1, len(nums)):
                currProduct *= nums[j]
                maxProduct = max(maxProduct, currProduct)
        
        return maxProduct

  

File: Data Structures & Algorithms/house-robber-ii/submission-0.cpp
class Solution {
public:
    int rob(vector<int>& nums) {

        if (nums.size() <= 2) return *max_element(nums.begin(), nums.end());

        return max(robLinear(nums, 0, nums.size()-1), robLinear(nums, 1, nums.size()));

    }

private:
    int robLinear(vector<int>& nums, int s, int e) {

        int prev1 = 0;
        int prev2 = 0;

        for (int i = s; i < e; i++) {
            int current = max(prev1, prev2 + nums[i]);
            prev2 = prev1;
            prev1 = current;
        }

        return prev1;
    }    
};
